pep-0724.rst

```diff
@@ -0,0 +1,306 @@
+    # list[T | None])
+    def has_no_nones(val: list[T | None]) -> TypeGuard[list[T]]: 
+        return None not in val
+
+* When stricter semantics are possible, the output type for the ``False`` or
+  ``NN`` case can be determined:
+
+.. code-block:: python
+
+    def is_str(val: str | int) -> TypeGuard[str]: # Stricter mode possible
+        return isinstance(val, str)
+
+    def func(val: str | int):
+        if is_str(val):
+            # Type checkers can assume val is a 'str' in this branch
+        else:
+            # Type checkers can assume val is an 'int' in this branch
+
+* When the output type of the ``TypeGuard`` is a union, the type checker can
+  apply additional type narrowing based on the type of the first input
+  argument, eliminating union elements that are impossible given the input
+  argument type:
+
+.. code-block:: python
+
+    def is_cardinal_direction(val: str) -> TypeGuard[Literal["N", "S", "E", "W"]]:
+        return val in ("N", "S", "E", "W")
+
+    def func(direction: Literal["NW", "E"]):
+        if is_cardinal_direction(direction):
+            # Literal["E"] The type cannot be "N", "S" or "W" here because of
+            # argument type
+        else:
+            # Literal["NW"]
+
+
+This logic can be summed up in a table:
+
+============ ======================= ===================
+|space|      Non strict TypeGuard    Strict TypeGuard
+============ ======================= ===================
+Applies when R not consistent with P R consistent with P
+NP is ..     :math:`R`               :math:`A \land R`
+NN is ..     :math:`A`               :math:`A \land \neg{R}`
+============ ======================= ===================
+
+Intersection of types and type negation are not defined and are left up to the
+type checker to decide on how to implement. Future `extensions`__ to the type
+system may change this behavior though.
+
+__ https://github.com/CarliJoy/intersection_examples/blob/main/specification.rst
+
+.. |space| unicode:: 0x2008
+
+Additional Examples
+===================
+
+Any
+---
+
+``Any`` is consistent [#isconsistent]_ with any other type, which means
+stricter semantics can be applied. 
+
+.. code-block:: python
+
+    def is_a(x: Any) -> TypeGuard[A]: # Stricter TypeGuard possible
+        return isinstance(x, A)
+
+    def test(x: A | B):
+        if is_a(x):
+            # x is of type A here
+        else:
+            # x is of type B here
+
+Degenerate case
+---------------
+
+It is possible to break the type checker by making a ``TypeGuard`` that is
+incorrect:
+
+.. code-block:: python
+
+    def is_positive_int(val: int | str) -> TypeGuard[int]:
+        return isinstance(val, int) and val > 0
+
+    def func(val: int | str):
+        if is_positive_int(val):
+            # Type checker assumes int here
+        else:
+            # Type checker assumes str incorrectly here
+
+As :pep:`647 <647#enforcing-strict-narrowing>` stated, users can create invalid
```

Comment:
Worth noting that under PEP 647, this particular TypeGuard is actually correct and meaningful. I still think that's worth calling out explicitly in the "Backwards Compatibility" section, though that section is now much better than when I last read the PEP.